runtime/src/iree/io/formats/irpa/irpa_builder.c

```diff
@@ -0,0 +1,413 @@
+            .pattern_length = target_entry.storage.splat.pattern_length,
+        };
+        memcpy(splat_entry.pattern, target_entry.storage.splat.pattern,
+               sizeof(splat_entry.pattern));
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0,
+            iree_io_stream_write(stream, sizeof(splat_entry), &splat_entry));
+        break;
+      }
+      case IREE_IO_PARAMETER_INDEX_ENTRY_STORAGE_TYPE_FILE: {
+        iree_io_parameter_archive_data_entry_t data_entry = {
+            .header =
+                {
+                    .entry_size = sizeof(data_entry),
+                    .type = IREE_IO_PARAMETER_ARCHIVE_ENTRY_TYPE_DATA,
+                    .flags = 0,
+                    .name = name_ref,
+                    .metadata = metadata_ref,
+                    .minimum_alignment =
+                        IREE_IO_PARAMETER_ARCHIVE_DEFAULT_DATA_ALIGNMENT,
+                },
+            .storage =
+                {
+                    .offset = target_entry.storage.file.offset,
+                    .length = target_entry.length,
+                },
+        };
+        target_entry.storage.file.handle = file_handle;
+        target_entry.storage.file.offset += storage_segment.offset;
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0, iree_io_stream_write(stream, sizeof(data_entry), &data_entry));
+        break;
+      }
+      default: {
+        IREE_RETURN_AND_END_ZONE_IF_ERROR(
+            z0, iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                                 "unhandled entry type %d",
+                                 (int)source_entry->type));
+      }
+    }
+
+    // Add the entry to the target_index referencing the location in the file
+    // reserved for the entry storage.
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_io_parameter_index_add(target_index, &target_entry));
+  }
+
+  // Write out the metadata table.
+  for (iree_host_size_t i = 0;
+       i < iree_io_parameter_index_count(builder->index); ++i) {
+    // Query the source entry template.
+    const iree_io_parameter_index_entry_t* source_entry = NULL;
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_io_parameter_index_get(builder->index, i, &source_entry));
+
+    // Write header metadata.
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_io_stream_write(stream, source_entry->key.size,
+                                 source_entry->key.data));
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_io_stream_write(stream, source_entry->metadata.data_length,
+                                 source_entry->metadata.data));
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_add_splat_entry(
+    iree_io_parameter_archive_builder_t* builder, iree_string_view_t name,
+    iree_const_byte_span_t metadata, const void* pattern,
+    uint8_t pattern_length, iree_io_physical_size_t data_length) {
```

Comment:
I think you need a check that pattern_length does not exceed the size of where it is being written.